### ISSUE TITLE
Remove unused OTUs (not present in any tree) from studies

### ISSUE THREAD

**Issue by jimallman** [OpenTreeOfLife/opentree#909]
I just realized, we're still not cleaning up unused OTUs when removing trees from a NexSON study. 

[Here's an extreme example](https://devtree.opentreeoflife.org/curator/study/edit/tt_35/?tab=otu-mapping), where the original (large) trees were removed and replaced with tiny trees. There are only a few OTUs in the new trees, but thousands are still stored in NexSON.

**Comment by kcranston** [comment]
In addition to fixing this in the curator app, should this also be part of the validation for phylesystem-api? Thinking ahead to allowing other clients to push to phylesystem. @mtholder ?

**Comment by mtholder** [comment]
It certainly would not be hard to write a function to purge unreferenced OTU elements.

It is a bit hard to say what the right behavior is in general and what the UI should be.

If someone:
  1.  has put in curation effort into mapping otu labels to OTT, 
  2.  figures out that they added the "wrong" tree

then they might move the study through a state of few/no trees before they add in the preferred trees.
In such a situation it would be irritating to have lost of otu mapping.

**Comment by jimallman** [comment]
In such a situation, would we be able to recognize the same (already-mapped) OTUs in a new tree? By matching labels exactly, perhaps? (Perhaps I'm mistaken, but I thought we didn't really try to consolidate "matching" OTUs across trees.)

**Comment by kcranston** [comment]
Another possibility might be to store *all* of our OTU mappings independently of studies. This would allow both 'hey! we've seen this OTU before and someone else previously mapped it to X" and might be of use for other projects.

**Comment by mtholder** [comment]
I think that we do look for exact matches to OTU labels whenever a new tree is added (otherwise we would get a new OTU element for every tree in which it occurs if the trees came in one by one in newick).

**Comment by mtholder** [comment]
(my last comment was in response to @jimallman in case that wasn't clear).

**Comment by jimallman** [comment]
Gotcha. So we do in effect preserve the labor of matching original labels (presumably shared across related trees) to the OT taxonomy. That does make culling unused OTUs a harder decision, perhaps one we should put to the user?

In any case, there could be an option for `peyotl.manip.merge_otus_and_trees` (which operates on complete NexSON documents) to cull unused OTUs, or not. We could call this after removing trees from a study, as well as adding them.

**Comment by mtholder** [comment]
If we added a new filter "in some trees" as the default choice of what OTUs to show, then these would not annoy the user. but they could be used later if a new trees. 

I'm sorry to always suggest a path forward that involves UI dev not back-end - I promise I'm not *just* being lazy ;-)

**Comment by jimallman** [comment]
In the example I posted above, the study is pretty slow to load and save, mainly due to the thousands of unused OTUs it carries. But if we're not worried about "clutter" in the NexSON file, client-side filtering does seem like an elegant solution. In fact, we already imply this in the curation app's OTU list filters:
- In all trees
- In preferred trees
- In non-preferred trees

I think my code is just lazy here, and "In all trees" just shows all OTUs in the study. I can change this and we're done!

**Comment by kcranston** [comment]
I think we should be a little worried about too much clutter in the
nexsons. Thousands of unused OTUs is a lot of clutter.

**Comment by jimallman** [comment]
Agreed, but the revised filter behavior (done in commit d0e1430 above) seems better, i.e. we preserve these OTUs but always hide them in the OTU mapping list.

Or we could add a list filter option "Not used in any tree", if we want to show them explicitly. This list view might also be a sensible place to put a "Remove unused OTUs" button, if we want to let them explicitly make that choice.

**Comment by jimallman** [comment]
> If we added a new filter "in some trees" as the default choice of what OTUs to show

Ah, perhaps my labeling for this filter is unclear. The current default _does_ exactly this (show all OTUs that appear in any tree), but maybe its name "In all trees" suggests that only an OTU that's in **every** tree would appear. Hmmm. Maybe "In any tree" would say this better?

**Comment by jimallman** [comment]
> Another possibility might be to store all of our OTU mappings independently of studies (in addition to storing relevant mappings within a nexson). This would allow both 'hey! we've seen this OTU before and someone else previously mapped it to X" and might be of use for other projects.

This is a neat idea! Maybe more of an OTI solution? A separate, fast index that checks for approved mappings of a given original label.

**Comment by jimallman** [comment]
FYI, the filter fix (hides unused OTUs in all cases) is on **devtree**.

**Comment by jimallman** [comment]
I'm marking this for further discussion, mainly around two questions:
- Should we offer an explicit option to purge all unused OTUs (not in any tree)?
- Should we pursue a more ambitious, [separate storage for already-mapped OTUs](https://github.com/OpenTreeOfLife/opentree/issues/909#issuecomment-186307719)?
